execution-engine/src/wasi/common.rs

```diff
@@ -0,0 +1,1519 @@
+    /// again from the global policy file.
+    program_arguments: Vec<String>,
+    /// The principal that accesses the filesystem. This information is used in path_open.
+    principal: Principal,
+    /// The exit code, if program calls proc_exit.
+    exit_code: Option<u32>,
+}
+
+/// A macro for lock the VFS in WASIWrapper and store the result in the variable,
+/// which will be captured by `$vfs` in the macro.
+/// It is assumed that `self` is a WASIWrapper.
+/// If the locks fails, it returns Busy error code.
+macro_rules! lock_vfs {
+    ($vfs:ident) => {
+        match $vfs.filesystem.lock() {
+            Ok(o) => o,
+            Err(_) => return ErrNo::Busy,
+        }
+    };
+}
+
+/// A macro to call try_from on a untyped raw wasi parameter `$var`.
+/// It converts `$var` to the typed version, `$t`, of the parameter as a value,
+/// or returns from the function with the `Inval` error code.
+macro_rules! decode_wasi_arg {
+    ($var:ident, $t:ty) => {
+        match <$t>::try_from($var) {
+            Err(_) => return ErrNo::Inval,
+            Ok(o) => o,
+        }
+    };
+}
+
+/// A macro to call read_cstring at `$address` of length `$len`.
+/// It returns String, or the error code.
+macro_rules! read_cstring {
+    ($mem:ident, $address:ident, $len:ident) => {
+        match $mem.read_cstring($address, $len) {
+            Ok(o) => o,
+            Err(e) => return e,
+        }
+    };
+}
+
+impl WASIWrapper {
+    /// The name of the WASM program's entry point.
+    pub(crate) const ENTRY_POINT_NAME: &'static str = "_start";
+    /// The name of the WASM program's linear memory.
+    pub(crate) const LINEAR_MEMORY_NAME: &'static str = "memory";
+    /// The name of the containing module for all WASI imports.
+    pub(crate) const WASI_SNAPSHOT_MODULE_NAME: &'static str = "wasi_snapshot_preview1";
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Creating and modifying runtime states.
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// Creates a new initial `WASIWrapper`.
+    pub fn new(filesystem: Arc<Mutex<FileSystem>>, principal: Principal) -> Self {
+        Self {
+            filesystem,
+            environment_variables: Vec::new(),
+            program_arguments: Vec::new(),
+            principal,
+            exit_code: None,
+        }
+    }
+
+    ///////////////////////////////////////////////////////
+    //// Functions for the execution engine internal
+    ///////////////////////////////////////////////////////
+
+    /// An internal function for the execution engine to directly read the file.
+    #[inline]
+    pub(crate) fn read_file_by_filename(&mut self, file_name: &str) -> Result<Vec<u8>, ErrNo> {
+        let mut fs = self.filesystem.lock().map_err(|_| ErrNo::Busy)?;
+        fs.read_file_by_filename(&Principal::InternalSuperUser, file_name)
+    }
+
+    pub(crate) fn exit_code(&self) -> Option<u32> {
```

Comment:
No `rustdoc` comment, and this can be inlined.  Also, I'm (somewhat) surprised that this compiles!  Shouldn't you be getting errors about trying to move out of a borrowed context here?  I'd have expected you'd have needed to do:

```rust
fn exit_code(&self) -> &Option<u32> {
  &self.exit_code
}
```

or

```rust
fn exit_code(&self) -> Option<&u32> {
  self.exit_code.as_ref()
}
```

Reply:
It can compile since u32 implements Copy trait. The rust implicitly calls copy.
[Here](https://doc.rust-lang.org/std/marker/trait.Copy.html) you can find a list of type that implements Copy trait. Those types will have _copy semantics_ when assignment happens rather than rust standard _move semantic_ .

Reply:
*The more you know!*